File: objc/objcfind/Tests/objcfindTests/FindOptionsTests.m
//
//  FindOptionsTests.m
//  objcfind_tests
//
//  Created by Cary Clark on 11/11/18.
//  Copyright © 2018 Cary Clark. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "FileUtil.h"
#import "FindOptions.h"

@interface FindOptionsTests : XCTestCase

@end

@implementation FindOptionsTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testSettingsFromMinimalArgs {
    NSError *error = nil;
    FindConfig *config = [[FindConfig alloc] init];
    FindOptions *options = [[FindOptions alloc] initWithConfig:config error:&error];
    NSArray *args =[NSArray arrayWithObjects:@"objfind",@".",nil];
    FindSettings *settings = [options settingsFromArgs:args error:&error];
    XCTAssert(![settings archivesOnly]);
    XCTAssert(![settings debug]);
    XCTAssert(![settings followSymlinks]);
    XCTAssert(![settings includeArchives]);
    XCTAssert(![settings includeHidden]);
    XCTAssert(![settings printDirs]);
    XCTAssert([settings printFiles]);
    XCTAssert(![settings printUsage]);
    XCTAssert(![settings printVersion]);
    XCTAssert(![settings verbose]);
    
    XCTAssert([[settings paths] count] == 1);
    XCTAssert([[[settings paths] objectAtIndex:0] isEqual:@"."]);
}

- (void)testSettingsFromValidArgs {
    NSError *error = nil;
    FindConfig *config = [[FindConfig alloc] init];
    FindOptions *options = [[FindOptions alloc] initWithConfig:config error:&error];
    NSArray *args =[NSArray arrayWithObjects:@"objfind",@"-x",@"java,scala",@".",nil];
    FindSettings *settings = [options settingsFromArgs:args error:&error];
    
    XCTAssert([[settings inExtensions] count] == 2);
    XCTAssert([[[settings inExtensions] objectAtIndex:0] isEqual:@"java"]);
    XCTAssert([[[settings inExtensions] objectAtIndex:1] isEqual:@"scala"]);
    XCTAssert([[settings paths] count] == 1);
    XCTAssert([[[settings paths] objectAtIndex:0] isEqual:@"."]);
}

- (void)testSettingsFromJson {
    NSString *startPath = @"~/src/xfind";
    NSString *json = [NSString stringWithFormat:@"{\n"
                      "\"path\": \"%@\",\n"
                      "\"in-ext\": [\"js\", \"ts\"],\n"
                      "\"out-dirpattern\": \"node_module\",\n"
                      "\"out-filepattern\": [\"temp\"],\n"
                      "\"debug\": true,\n"
                      "\"followsymlinks\": true,\n"
                      "\"includehidden\": true\n"
                      "}", startPath];

    NSData *data = [json dataUsingEncoding:NSUTF8StringEncoding];

    NSError *error = nil;
    FindConfig *config = [[FindConfig alloc] init];
    FindOptions *options = [[FindOptions alloc] initWithConfig:config error:&error];
    FindSettings *settings = [options settingsFromData:data error:&error];

    XCTAssert([[settings inExtensions] count] == 2);
    XCTAssert([[[settings inExtensions] objectAtIndex:0] isEqual:@"js"]);
    XCTAssert([[[settings inExtensions] objectAtIndex:1] isEqual:@"ts"]);
    XCTAssert([[settings outDirPatterns] count] == 1);
    XCTAssert([[[[settings outDirPatterns] objectAtIndex:0] pattern] isEqual:@"node_module"]);
    XCTAssert([[settings outFilePatterns] count] == 1);
    XCTAssert([[[[settings outFilePatterns] objectAtIndex:0] pattern] isEqual:@"temp"]);
    XCTAssert([[settings paths] count] == 1);
    XCTAssert([[[settings paths] objectAtIndex:0] isEqual:@"~/src/xfind"]);
    XCTAssert([settings debug]);
    XCTAssert([settings verbose]);
    XCTAssert([settings followSymlinks]);
    XCTAssert([settings includeHidden]);
}

@end

File: swift/swiftfind/Tests/swiftfindTests/FindOptionsTests.swift
//
//  FindOptionsTests.swift
//  swiftfind
//
//  Created by Cary Clark on 5/18/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import XCTest

import swiftfind

class FindOptionsTests: XCTestCase {
    let options = FindOptions()
    let startPath: String = "."
    var requiredArgs: [String] = []

    override func setUp() {
        super.setUp()
        requiredArgs = [startPath]
    }

//    override func tearDown() {
//        super.tearDown()
//    }

    func testSettingsEqualDefaultFindSettings() {
        let settings: FindSettings = try! options.settingsFromArgs(requiredArgs)
        XCTAssert(settings.archivesOnly == DefaultFindSettings.archivesOnly, "archivesOnly == false")
        XCTAssert(settings.debug == DefaultFindSettings.debug, "debug == false")
        XCTAssert(settings.followSymlinks == DefaultFindSettings.followSymlinks, "followSymlinks == false")
        XCTAssert(settings.includeHidden == DefaultFindSettings.includeHidden, "includeHidden == false")
        XCTAssert(settings.printDirs == DefaultFindSettings.printDirs, "printDirs == false")
        XCTAssert(settings.printFiles == true, "printFiles == true")
        XCTAssert(settings.printUsage == DefaultFindSettings.printUsage, "printUsage == false")
        XCTAssert(settings.printVersion == DefaultFindSettings.printVersion, "printVersion == false")
        XCTAssert(settings.includeArchives == DefaultFindSettings.includeArchives,
                  "includeArchives == false")
        XCTAssert(settings.verbose == DefaultFindSettings.verbose, "verbose == false")
    }

    func testSettingsFromArgs() {
        let otherArgs: [String] = ["--in-ext", "scala,swift", "--debug"]
        let settings: FindSettings = try! options.settingsFromArgs(requiredArgs + otherArgs)
        print("settings: \(settings)")
        XCTAssert(settings.debug, "debug == true")
        XCTAssert(settings.verbose, "verbose == true")
        XCTAssertEqual(2, settings.inExtensions.count)
        XCTAssertTrue(settings.inExtensions.contains("scala"))
        XCTAssertTrue(settings.inExtensions.contains("swift"))
        XCTAssertEqual(1, settings.paths.count)
        XCTAssertTrue(settings.paths.contains(startPath))
    }

    func testSettingsFromJson() {
        let jsonString = """
        {
          "path": "~/src/xfind/",
          "in-ext": ["js", "ts"],
          "out-dirpattern": ["_", "ansible", "bak", "build", "chef", "node_module", "target", "test", "typings"],
          "out-filepattern": ["gulpfile", ".min."],
          "debug": true,
          "followsymlinks": true,
          "includehidden": true,
          "printdirs": true,
          "printfiles": true
        }
        """
        let settings = try! options.settingsFromJson(jsonString)
        print("settings: \(settings)")
        XCTAssertTrue(settings.debug, "debug == true")
        XCTAssertTrue(settings.followSymlinks, "followSymlinks == true")
        XCTAssertTrue(settings.includeHidden, "includeHidden == true")
        XCTAssertEqual(2, settings.inExtensions.count)
        XCTAssertTrue(settings.inExtensions.contains("js"))
        XCTAssertTrue(settings.inExtensions.contains("ts"))
        XCTAssertEqual(9, settings.outDirPatterns.count)
        XCTAssertEqual(1, settings.paths.count)
        XCTAssertTrue(settings.paths.contains("~/src/xfind/"))
        XCTAssertTrue(settings.printDirs)
        XCTAssertTrue(settings.printFiles)
        XCTAssertTrue(settings.verbose, "verbose == true")
    }

    static var allTests = [
        ("testSettingsEqualDefaultFindSettings", testSettingsEqualDefaultFindSettings),
        ("testSettingsFromArgs", testSettingsFromArgs),
        ("testSettingsFromJson", testSettingsFromJson)
    ]
}
